File: core/typescript/src/registry.ts
/**
 * The frozen suite registry (spec §4.2, docs/09 §6).
 *
 * A hard-coded table, not a plugin surface: adding a suite is a code change in
 * every core plus vectors plus a spec revision. Two suites are *registered*
 * (recognized by `isCiphertext`, spec §3.4); only one is *implemented* by this
 * core.
 *
 * 0xFF02 is registered and deliberately unbuilt: its AEAD (XChaCha20-Poly1305)
 * has no citable normative definition (spec §4.2 note, gap G7). Registering it
 * keeps recognition correct -- a 0xFF02 envelope is ciphertext, never
 * unmigrated plaintext, which is what §3.4's double-encryption story is about
 * -- while `implemented: false` keeps it off every allow-list this core will
 * accept at construction (see config.ts for the reasoning).
 */

export const FMT_VER = 0x01;

/** Header bytes that precede the nonce: fmt_ver(1) + suite_id(2) + key_id(16) + msg_seed(32). */
export const HEADER_FIXED_LEN = 51;
export const KEY_ID_LEN = 16;
export const MSG_SEED_LEN = 32;

export interface Suite {
  readonly id: number;
  readonly name: string;
  readonly aead: "AES-256-GCM" | "XChaCha20-Poly1305";
  readonly keyLen: number;
  readonly nonceLen: number;
  readonly tagLen: number;
  readonly commitLen: number;
  readonly kdf: "HKDF-SHA-512";
  /** Whether this core can perform the suite's AEAD. */
  readonly implemented: boolean;
  /** Spec §4.8: identifiers in 0xFF00-0xFFFF are provisional and require arming to write. */
  readonly provisional: boolean;
}

export const SUITE_FF01 = 0xff01;
export const SUITE_FF02 = 0xff02;

const SUITES: ReadonlyMap<number, Suite> = new Map<number, Suite>([
  [
    SUITE_FF01,
    {
      id: SUITE_FF01,
      name: "FLE-AES256GCM-HKDF-SHA512-PROVISIONAL",
      aead: "AES-256-GCM",
      keyLen: 32,
      nonceLen: 12,
      tagLen: 16,
      commitLen: 32,
      kdf: "HKDF-SHA-512",
      implemented: true,
      provisional: true,
    },
  ],
  [
    SUITE_FF02,
    {
      id: SUITE_FF02,
      name: "FLE-XCHACHA20POLY1305-HKDF-SHA512-PROVISIONAL",
      aead: "XChaCha20-Poly1305",
      keyLen: 32,
      nonceLen: 24,
      tagLen: 16,
      commitLen: 32,
      kdf: "HKDF-SHA-512",
      implemented: false,
      provisional: true,
    },
  ],
]);

export function getSuite(id: number): Suite | undefined {
  return SUITES.get(id);
}

export function isRegistered(id: number): boolean {
  return SUITES.has(id);
}

export function isProvisionalId(id: number): boolean {
  // §4.8: one masked comparison on suite_id.
  return (id & 0xff00) === 0xff00;
}

/** Fixed (plaintext-independent) envelope overhead for a suite, in bytes. */
export function fixedOverhead(s: Suite): number {
  return HEADER_FIXED_LEN + s.nonceLen + s.tagLen + s.commitLen;
}

/** The smallest envelope any registered suite can produce (an empty plaintext). */
export function minRegisteredEnvelopeLen(): number {
  let min = Number.POSITIVE_INFINITY;
  for (const s of SUITES.values()) min = Math.min(min, fixedOverhead(s));
  return min;
}

export function registeredSuiteIds(): number[] {
  return [...SUITES.keys()];
}

File: core/java/fieldseal-core/src/main/java/dev/fieldseal/core/internal/registry/Registry.java
package dev.fieldseal.core.internal.registry;

import java.util.List;
import java.util.Optional;

/**
 * The frozen suite table (spec §4.2; docs/09 §6). There is deliberately no registration API:
 * adding a suite is a code change in every core, plus vectors, plus a spec revision (docs/09 §6).
 *
 * <p>{@code 0xFF01} is built. {@code 0xFF02} is registered and unbuilt (docs/27 §10): it is
 * recognized as an envelope, because recognition follows the registry and not what this core can
 * perform (spec §3.4), but no client here can decrypt or write under it. {@code 0x0001} and
 * {@code 0x0002} are reserved and unassigned (spec §4.2), so they are not in the table and an
 * envelope naming them is not recognized.
 */
public final class Registry {

    public static final Suite FF01 = new Suite(0xFF01, "FLE-AES256GCM-HKDF-SHA512-PROVISIONAL",
            "AES-256-GCM", 32, 12, 16, 32, true);

    public static final Suite FF02 = new Suite(0xFF02,
            "FLE-XCHACHA20POLY1305-HKDF-SHA512-PROVISIONAL", "XChaCha20-Poly1305", 32, 24, 16, 32,
            false);

    private static final List<Suite> ALL = List.of(FF01, FF02);

    private Registry() {}

    /** Every registered suite, in {@code suite_id} order. */
    public static List<Suite> all() {
        return ALL;
    }

    public static Optional<Suite> lookup(int suiteId) {
        for (Suite s : ALL) {
            if (s.id() == suiteId) {
                return Optional.of(s);
            }
        }
        return Optional.empty();
    }
}
